libafl/src/stages/mod.rs

```diff
@@ -122,51 +111,75 @@ where
     E: UsesState<State = S>,
     EM: UsesState<State = S>,
     Z: UsesState<State = S>,
-    S: UsesInput,
+    S: UsesInput + HasStageStatus,
 {
     fn perform_all(
         &mut self,
         _: &mut Z,
         _: &mut E,
-        _: &mut S,
+        stage: &mut S,
         _: &mut EM,
-        _: CorpusId,
     ) -> Result<(), Error> {
-        Ok(())
+        if stage.current_stage()?.is_some() {
+            Err(Error::illegal_state(
+                "Got to the end of the tuple without completing resume.",
+            ))
+        } else {
+            Ok(())
+        }
     }
 }
 
 impl<Head, Tail, E, EM, Z> StagesTuple<E, EM, Head::State, Z> for (Head, Tail)
 where
     Head: Stage<E, EM, Z>,
-    Tail: StagesTuple<E, EM, Head::State, Z>,
+    Tail: StagesTuple<E, EM, Head::State, Z> + HasConstLen,
     E: UsesState<State = Head::State>,
     EM: UsesState<State = Head::State>,
     Z: UsesState<State = Head::State>,
+    Head::State: HasStageStatus,
 {
     fn perform_all(
         &mut self,
         fuzzer: &mut Z,
         executor: &mut E,
         state: &mut Head::State,
         manager: &mut EM,
-        corpus_idx: CorpusId,
     ) -> Result<(), Error> {
-        // Perform the current stage
-        self.0
-            .perform(fuzzer, executor, state, manager, corpus_idx)?;
+        match state.current_stage()? {
+            Some(idx) if idx < Self::LEN => {
+                // do nothing; we are resuming
+            }
+            Some(idx) if idx == Self::LEN => {
+                // perform the stage, but don't set it
+                Head::Progress::initialize_progress(state)?;
+                self.0.perform(fuzzer, executor, state, manager)?;
+                Head::Progress::clear_progress(state)?;
+                state.clear_stage()?;
+            }
+            Some(idx) if idx > Self::LEN => {
+                unreachable!("We should clear the stage index before we get here...");
+            }
+            // this is None, but the match can't deduce that
+            _ => {
+                state.set_stage(Self::LEN)?;
```

Comment:
In this case, yes. That's why the indexes are backwards on the resuming implementation.

Reply:
> That's why the indexes are backwards on the resuming implementation.

which part of the code?
i was about to say that you are doing it forward here
```
            Some(idx) if idx < Self::LEN => {
                // do nothing; we are resuming
            }
            Some(idx) if idx == Self::LEN => {
                // perform the stage, but don't set it
                Head::Progress::initialize_progress(state)?;
                self.0.perform(fuzzer, executor, state, manager)?;
                Head::Progress::clear_progress(state)?;
                state.clear_stage()?;
            }
```

Reply:
idx represents the _destination_ index; `LEN` is the counter for the current position. The first branch can be read as, "if the targeted index is less than the current index", but since the indices are in reverse order this works.

Reply:
ah i see.